lib/course.js

```diff
@@ -33,12 +33,44 @@ async function getCourseLessons(course) {
   )
 }
 
-export async function getCourse(course_id) {
-  const docRef = doc(db, 'courses', course_id)
-  const courseDoc = await getDoc(docRef)
-  const course = { id: course_id, ...courseDoc.data() }
-  const lessons = await getCourseLessons(course)
-  return { ...course, lessons }
+async function getCourseById(db, courseId) {
+  const docRef = doc(db, 'courses', courseId);
+  const docSnap = await getDoc(docRef);
+
+  if (!docSnap.exists()) {
+    throw new Error("Course not found!");
+  }
+
+  const data = { id: docSnap.id, ...docSnap.data() };
+
+  const lessons = await getCourseLessons(data);
+  return { ...data, lessons };
+}
+
+async function getGroupBySlug(db, groupSlug) {
+  const colRef = collection(db, 'study_groups');
+  const q = query(colRef, where('slug', '==', groupSlug));
+  const querySnapshot = await getDocs(q);
+
+  if (querySnapshot.empty) {
+    throw new Error("Group not found!");
+  }
+
+  const docSnap = querySnapshot.docs[0];
+  return { id: docSnap.id, ...docSnap.data() };
+}
+
+const kindToFunctionMap = {
+  course: getCourseById,
+  group: getGroupBySlug
+};
+
+export async function getCourse(kind, identifier) {
```

Comment:
we should rename this. `Study Groups` are not courses. I understand you intention, reusing code, but this signature confuses me. Also the implementation doesn't need this `kindToFunctionMap`. If can directly get the function from the kind. My suggestion:

```javascript
function getCourse(courseId) { }
function getStudyGroup(slug) { }

function getDocument(collection, id) { 
   return self["get" + collection](id)
} 

getDocument("StudyGroup", "DAO")
getDocument("Course", "Solidity")